plugins/modules/zypper_repository.py

```diff
@@ -357,18 +357,21 @@ def main():
         'priority': module.params['priority'],
     }
     # rewrite bools in the language that zypper lr -x provides for easier comparison
-    if module.params['enabled']:
-        repodata['enabled'] = '1'
-    else:
-        repodata['enabled'] = '0'
-    if module.params['disable_gpg_check']:
-        repodata['gpgcheck'] = '0'
-    else:
-        repodata['gpgcheck'] = '1'
-    if module.params['autorefresh']:
-        repodata['autorefresh'] = '1'
-    else:
-        repodata['autorefresh'] = '0'
+    if 'enabled' in module.params:
+        if module.params['enabled']:
+            repodata['enabled'] = '1'
+        else:
+            repodata['enabled'] = '0'
+    if 'disable_gpg_check' in module.params:
+        if module.params['disable_gpg_check']:
+            repodata['gpgcheck'] = '0'
+        else:
+            repodata['gpgcheck'] = '1'
+    if 'autorefresh' in module.params:
+        if module.params['autorefresh']:
+            repodata['autorefresh'] = '1'
+        else:
+            repodata['autorefresh'] = '0'
```

Comment:
`module.params` always contains **all possible module options**, so `'xxx' in module.params` will always evaluate to `True` (if `xxx` is a valid module option). Also these three options have defaults, so even if the user did not provide them, they will always be set to their default value in `module.params`.

If you want to test whether the values have been provided by the user or not, you need to remove the defaults, and check for `None` instead (like `if module.params['autorefresh'] is not None:`).

Reply:
Interesting. I thought, I had tested it locally and adding the condition works. I'll give this another try and I'll also try to add some tests

Reply:
On the other hand I think this might break the logic where the plugin compares the configuration passed via module.params against what is found on the file system - not passing a value might result in the code detecting a change in configuration....

Also: is this the default behaviour in ansible-modules: not passing a parameter that has a default configured will change an existing configuration that was configured against the default value? That's something, I wouldn't expect

Reply:
> Also: is this the default behaviour in ansible-modules: not passing a parameter that has a default configured will change an existing configuration that was configured against the default value? That's something, I wouldn't expect

That depends on whether options have a default value or not. Having a default for options that are used for idempotency checks have this side-effect, which is usually not what the user wants. That's why it's best to only have default values for options that are not used for idempotency check, but never have defaults for options that are used for idempotency checks. (Unless there are very good reasons for a specific option; which can happen, but is not very likely.)

To fix this, it's usually best to remove the default values, and change the code so that if the user does not provide the option, the old default value is used when something is created.

Also please note that you are changing the behavior of the module, as I understand it, which is a breaking change. You need to control the behavior by a new option so that existing tasks **will behave exactly as before**.

Reply:
> `module.params` always contains **all possible module options**, so `'xxx' in module.params` will always evaluate to `True` (if `xxx` is a valid module option). Also these three options have defaults, so even if the user did not provide them, they will always be set to their default value in `module.params`.

I beg to differ, @felixfontein . If the value of the parameter is `None`, then that stmt is equivalent to `if None:`, which evaluates to `False`. He's not checking whether the parameter exists, he is using the parameter value, and `None` is... `False`-ish.

Reply:
The statement `if 'xxx' in module.params:` is **not** equivalent to `if None` if `module.params['xxx']` is `None`.

Reply:
You're right of course. I was mislead by the fact that the lines of code selected for this comment do not include that part. Re-reading I see I missed some bits, and I have even submitted a suggestion fixing that up.